tests/ConnectDisconnectTest.test.nut

```diff
@@ -0,0 +1,211 @@
+        assertTrue(!server.isconnected(), "should NOT be connected!");
+        assertTrue(!cm.isConnected(), "CM should NOT report connected state!");
+        cm.connect();
+      }.bindenv(this));
+
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+      assertTrue(!cm.isConnected(), "CM should NOT report connected state!");
+
+      cm.onConnect(function() {
+        resolve();
+      }.bindenv(this));
+
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+      assertTrue(!cm.isConnected(), "CM should NOT report connected state!");
+    }.bindenv(this))
+    .then(function(val) {
+        assertTrue(cm.isConnected(), "CM should report state as connected!");
+        assertTrue(server.isconnected(), "should be connected again!");
+        _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+        _resetCM.call(this);
+        throw valueOrReason;
+      }.bindenv(this));
+  }
+
+/*
+* Verifies that more than one connect() invocation doesn't lead to disconnection
+*
+*/
+  function testDoubleConnectAsync() {
+    assertTrue(server.isconnected(), "Failed pre-run saanity check");
+    return Promise(function(resolve, reject) {
+      cm.connect();
+      assertTrue(server.isconnected(), "should be connected!");
+      assertTrue(cm.isConnected(), "CM should report connected state!");
+
+      cm.onConnect(function() {
+        assertTrue(server.isconnected(), "should be connected!");
+        assertTrue(cm.isConnected(), "CM should report connected state!");
+        resolve();
+      }.bindenv(this));
+
+      cm.connect();
+      assertTrue(server.isconnected(), "should be connected!");
+      assertTrue(cm.isConnected(), "CM should report connected state!");
+    }.bindenv(this))
+    .then(function(val) {
+        assertTrue(cm.isConnected(), "CM should report state as connected!");
+        assertTrue(server.isconnected(), "should be connected again!");
+        _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+        _resetCM.call(this);
+        throw valueOrReason;
+      }.bindenv(this));
+  }
+
+  function _checkConnectForCommon(needDisconnect = false) {
+    local flag = 0;
+    assertTrue(server.isconnected(), "Failed pre-run saanity check");
+    if (needDisconnect) {
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+    }
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        assertTrue(!cm.isConnected(), "should NOT be connected!");
+        assertTrue(!server.isconnected(), "should NOT be connected!");
+
+        cm.onConnect(function() {
+          assertTrue(flag==2, "flag should be 2 now");
+          flag++;
+          resolve();
+        }.bindenv(this));
+        flag++;
+        cm.connect();
+      }.bindenv(this));
+
+      cm.connectFor(function() {
```

Comment:
As I understand connectFor and onNextConnect uses the same _queue. Do yo have some tests on cooperation between these methods ?